### ISSUE TITLE
Save / restore window state

### ISSUE THREAD

**Issue by quasado** [electron/electron#526]
Is there any API to save/restore the window's size / state / position? If not, what's the advised approach to do so?

**Comment by Imskull** [comment]
For anyone needs.
In your init js:

```
app.on('ready', function() {
  var path = require("path");
  var fs = require("fs");
  var initPath = path.join(app.getDataPath(), "init.json");
  var data;
  try {
    data = JSON.parse(fs.readFileSync(initPath, 'utf8'));
  }
  catch(e) {
  }
  // Create the browser window.
  mainWindow = new BrowserWindow((data && data.bounds) ? data.bounds : {width: 800, height: 600});
  ...
  mainWindow.on("close", function() {
    var data = {
      bounds: mainWindow.getBounds()
    };
    fs.writeFileSync(initPath, JSON.stringify(data));
  }
  ...
}
```
I couldn't find any way to use localStorage in the init js, have to use "fs" module instead.

**Comment by zeke** [comment]
Thanks for the snippet, @imskull. I used `electron-config` to persist the bounds to disk. Here's what I ended up with:

```js
const {app, BrowserWindow} = require('electron')
const Config = require('electron-config')
const config = new Config()
let win

app.on('ready', () => {
  let opts = {show: false}
  Object.assign(opts, config.get('winBounds'))
  win = new BrowserWindow(opts)
  win.loadURL(`file://${__dirname}/app/index.html`)

  win.once('ready-to-show', win.show)

  // save window size and position
  win.on('close', () => {
    config.set('winBounds', win.getBounds())
  })
})
```

**Comment by steverandy** [comment]
@zeke Is it safe to do fs operation during close event?
If user quits the app, how do we make sure file has been written?

**Comment by MarshallOfSound** [comment]
@steverandy Doing synchronous `fs` operations in the close event will work.  Asynchronous operations are not guaranteed to be completed.

**Comment by zeke** [comment]
> safe to do fs operation during close event?

Speaking of saving data when the app is closed: I used `electron-config` instead of localStorage because localStorage only saves if the app exits "cleanly". So if you kill your process (say with `^C`) you're out of luck.

See https://github.com/electron/electron/issues/1731#issuecomment-221922779

**Comment by andymartinwork** [comment]
Just for future visitors to this bug

The code above does not take into account if the app is 

* saved on an external monitor
* quit
* external monitor disconnected
* app opened again

You can quite often end up with the app being off the end of the screen (dependent on zoom factor). So something more complicated is necessary for a professional implementation.

**Comment by whyboris** [comment]
For reference, the documentation for `getBounds()` and `setBounds()` from Electron:
https://github.com/electron/electron/blob/master/docs/api/browser-window.md#winsetboundsbounds-animate

**Comment by zeke** [comment]
The [electron-window-state](https://ghub.io/electron-window-state) module is also a good option for this.

**Comment by sertxudeveloper** [comment]
> Just for future visitors to this bug
> 
> The code above does not take into account if the app is
> 
> * saved on an external monitor
> * quit
> * external monitor disconnected
> * app opened again
> 
> You can quite often end up with the app being off the end of the screen (dependent on zoom factor). So something more complicated is necessary for a professional implementation.

I solved this problem with this code:
```js
let window = {x: 267, y: 195} // Saved previous window position
let screen = electron.screen.getDisplayNearestPoint({ x: window.x, y: window.y })
if (!(window.x > screen.bounds.x && window.x < screen.size.width) || !(window.y > screen.bounds.y && window.y < screen.size.height)) {
    // The requested position cannot be accesible so the nearest point is applied
    window.x = screen.bounds.x
    window.y = screen.bounds.y
}
```

**Comment by simonbuchan** [comment]
The above code has a couple of bugs, e.g. comparing `window.x` to just `.width`, instead of `.x + .width`, using `screen.bounds` instead of `.workArea`, and not considering window size, and it will place the window at the top-left corner, which is a little ugly.

I've had success with this code to restore the position:

```
const options = {
  title: "My Saved Window",
  ...
};

const bounds = getSavedBounds();
if (bounds) {
  const area = electron.screen.getDisplayMatching(bounds).workArea;
  // If the saved position still valid (the window is entirely inside the display area), use it.
  if (
    bounds.x >= area.x &&
    bounds.y >= area.y &&
    bounds.x + bounds.width <= area.x + area.width &&
    bounds.y + bounds.height <= area.y + area.height
  ) {
    options.x = bounds.x;
    options.y = bounds.y;
  }
  // If the saved size is still valid, use it.
  if (bounds.width <= area.width || bounds.height <= area.height) {
    options.width = bounds.width;
    options.height = bounds.height;
  }
}

const window = new BrowserWindow(options);
```

There are other options, but this one is pretty safe and clear.

To save, I debounce the `move` and `resize` events:

```
window.on("resize", saveBoundsSoon);
window.on("move", saveBoundsSoon);
let saveBoundsCookie;

function saveBoundsSoon() {
  if (saveBoundsCookie) clearTimeout(saveBoundsCookie);
  saveBoundsCookie = setTimeout(() => {
    saveBoundsCookie = undefined;
    state.bounds = window.getNormalBounds();
    fs.writeFileSync(statePath, JSON.stringify(state, null, 2));
  }, 1000);
}
```

Note the use of `window.getNormalBounds()`, so you're not saving the maximized bounds.

**Comment by jayenashar** [comment]
How does Developer Tools remember its dock location and size?

**Comment by zcbenz** [comment]
I'm reopening this issue since this has proven to be an actually quite popular request and lots of apps have been implementing their own solutions.

**Comment by Tetrax-10** [comment]
above codes are buggy and complex, so i implemented this after a lot of testing this is the most accurate and simple solution to remember window properties.
i used `electron-store` to store values, you can use any storage methods you like.

**main.js**
```js
let windowConfig = {
    width: 1280,
    height: 680,
}

function createWindow() {
    Object.assign(windowConfig, CONFIG.get("winBounds"))
    mainWindow = new BrowserWindow(windowConfig)
    if (windowConfig.isMaximized) {
        mainWindow.maximize()
    }
    mainWindow.on("close", () => {
        Object.assign(windowConfig, {
            isMaximized: mainWindow.isMaximized()
        }, mainWindow.getNormalBounds())
        CONFIG.set("winBounds", windowConfig) // saves window's properties using electron-store
    })
}
```

**Comment by jerothor** [comment]
thanks to @Tetrax-10 , using that nifty `electron-store` lib:

```js
const { app, BrowserWindow } = require('electron')
const path = require('path')
const Store = require('electron-store')
const store = new Store()

function createWindow() {
    const win = new BrowserWindow({
        webPreferences: {
            preload: path.join(__dirname, 'preload.js'),
        }
    })

    win.setBounds(store.get('bounds'))

    win.on('close', () => {
        store.set('bounds', win.getBounds())
    })

    win.loadFile('index.html')
}
```

it even works for multiple monitors!

**Comment by nicholaszuccarelli** [comment]
@jerothor Perfect solution. Works a treat!

**Comment by adham667** [comment]
it says on your Google Summer of Code 2024 Ideas List that C++ is required to solve this issue isn't typescript sufficient to solve the problem ?

**Comment by kkocdko** [comment]
I wrote a post to summarize some cautions of window state restore. If someone want to implement it, consider reading this to check the corner cases.

https://kkocdko.site/post/202409161747

**Comment by felixrieseberg** [comment]
Hey everyone! Thank you all so much for giving Electron a try and thinking about ways to make it better. As you can probably tell from how this issue has been open for more than ten years, so far it has never been important enough for someone to close it directly or to otherwise make the resources available for it to happen. 

If you're wondering what you can do to get this into Electron, great news - Electron is truly open source! We, the maintainers, [welcome outside contributions - especially from individuals and companies interested in joining us in maintaining this platform](https://www.electronjs.org/governance).

If you're reading this and are thinking to yourself "Alright, I'll do it!" - good for you! For smaller changes, you can usually open a well-tested and documented PR directly on this repository. For larger changes, I'd recommend that you start with a "[request for comments](https://www.electronjs.org/blog/rfcs)". This will ensure that you don't spend a lot of time on a larger change without getting feedback about your approach first. Alternatively, you could always hire a consultant, either from the maintainer community our outside, to perform the work that's important to you.

In short, I'm closing this issue because we're currently _not planning_ this work, but if you care deeply about this issue, we'd welcome your contribution.

**Comment by citelao** [comment]
Adding onto the list of caveats for anyone who delves into this, while I was on the Windows team at Microsoft, I worked with @ekoschik to open-source some of his [guidelines and sample code on remembering window positions](https://github.com/microsoft/Windows-classic-samples/pull/416) on Windows (including an extensive [README](https://github.com/citelao/Windows-classic-samples/blob/93c2244a070b34ffc5c71408f80a7116695464b0/Samples/WindowPlacement/README.md), also [further some further revisions](https://github.com/citelao/Windows-classic-samples/pull/1/changes)). Unfortunately, I was not able to get the PR into a state I was comfortable with before I left, and my priorities are different at my new job 😄 .

I hope this helps, and thank you to anyone who moves this forward!

**Comment by wesleydarnell** [comment]
I dream of official support for this
